mongodb/node-mongodb-native
# feat(NODE-6275): Add CSOT support to GridFS

### Description

#### What is changing?

##### `GridFSUploadStream`
* Add CSOT support by piping `TimeoutContext` down through all operations over lifetime of `GridFSUploadStream`
* Migrate `checkIndexes` call from constructor to async `construct()` method to accomodate CSOT error handling

##### `GridFSDownloadStream`
* Add CSOT support by piping `TimeoutContext` down through all operations over lifetime of `GridFSDownloadStream`

##### `GridFSBucket`
* Add support for `timeoutMS` option to:
	* `GridFSBucket.openUploadStream()`
	* `GridFSBucket.openUploadStreamWithId()`
	* `GridFSBucket.openDownloadStream()`
	* `GridFSBucket.openDownloadStreamByName()`
	* `GridFSBucket.find()`
	* `GridFSBucket.drop()`
	* `GridFSBucket.rename()`
	* `GridFSBucket.delete()`
	
##### Testing
* Unskip CSOT gridfs spec tests
* Implement Prose test [6](https://github.com/mongodb/specifications/tree/master/source/client-side-operations-timeout/tests#6-gridfs---upload) and [7](https://github.com/mongodb/specifications/tree/master/source/client-side-operations-timeout/tests#7-gridfs---download) 
* Add int tests that verify that non-spec-required methods (`GridFSBucket.openUploadStreamWithId()` and  `GridFSBucket.openDownloadStreamByName()`) properly handle `timeoutMS`
* Add int test verifying that `'index'` event is emitted exactly once by `GridFSBucket` when creating upload streams

##### Is there new documentation needed for these changes?

Yes, punt to NODE-5688, also see NODE-6275 for specifics.

#### What is the motivation for this change?

NODE-6275






### Release Highlight




### Double check the following

- [X] Ran `npm run check:lint` script
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
